dali/operators/nvcvop/nvcvop.cc

```diff
@@ -225,13 +225,68 @@ nvcv::Tensor AsTensor(void *data, const TensorShape<> shape, DALIDataType daliDT
   return nvcv::TensorWrapData(inData);
 }
 
-void PushTensorsToBatch(nvcv::TensorBatch &batch, const TensorList<GPUBackend> &t_list,
-                        TensorLayout layout) {
-  for (int s = 0; s < t_list.num_samples(); ++s) {
-    batch.pushBack(AsTensor(t_list[s], layout));
+nvcv::Tensor AsTensor(const void *data, span<const int64_t> shape_data, const nvcv::DataType &dtype,
+                      const nvcv::TensorLayout &layout) {
+  int ndim = shape_data.size();
+  nvcv::TensorDataStridedCuda::Buffer inBuf;
+  inBuf.basePtr = static_cast<NVCVByte *>(const_cast<void *>(data));
+  inBuf.strides[ndim - 1] = dtype.strideBytes();
+  for (int d = ndim - 2; d >= 0; --d) {
+    inBuf.strides[d] = shape_data[d + 1] * inBuf.strides[d + 1];
+  }
+  nvcv::TensorShape out_shape(shape_data.data(), ndim, layout);
+  nvcv::TensorDataStridedCuda inData(out_shape, dtype, inBuf);
+  return nvcv::TensorWrapData(inData);
+}
+
+int64_t calc_num_frames(const TensorShape<> &shape, int first_spatial_dim) {
+  return (first_spatial_dim > 0) ?
+          volume(&shape[0], &shape[first_spatial_dim]) :
+          1;
+}
+
+void PushFramesToBatch(nvcv::TensorBatch &batch, const TensorList<GPUBackend> &t_list,
+                       int first_spatial_dim, int64_t starting_sample, int64_t frame_offset,
+                       int64_t num_frames, const TensorLayout &layout) {
+  int ndim = layout.ndim();
+  auto nvcv_layout = nvcv::TensorLayout(layout.c_str());
+  auto dtype = GetDataType(t_list.type());
+
+  std::vector<nvcv::Tensor> tensors;
+  tensors.reserve(num_frames);
+
+  const auto &input_shape = t_list.shape();
+  int64_t sample_id = starting_sample - 1;
+  auto type_size = dtype.strideBytes();
+  std::vector<int64_t> frame_shape(ndim, 1);
+
+  auto frame_stride = 0;
+  int sample_nframes = 0;
+  const uint8_t *data = nullptr;
+
+  for (int64_t i = 0; i < num_frames; ++i) {
+    if (frame_offset == sample_nframes) {
+      frame_offset = 0;
+      do {
+        ++sample_id;
+        auto sample_shape = input_shape[sample_id];
+        DALI_ENFORCE(sample_id < t_list.num_samples());
+        std::copy(&sample_shape[first_spatial_dim], &sample_shape[input_shape.sample_dim()],
+                  frame_shape.begin());
+        frame_stride = volume(frame_shape) * type_size;
+        sample_nframes = calc_num_frames(sample_shape, first_spatial_dim);
+      } while (sample_nframes * frame_stride == 0);  // we skip empty samples
+      data =
+          static_cast<const uint8_t *>(t_list.raw_tensor(sample_id)) + frame_stride * frame_offset;
+    }
+    tensors.push_back(AsTensor(data, make_span(frame_shape), dtype, nvcv_layout));
+    data += frame_stride;
+    frame_offset++;
```

Comment:
I think that combining the two loops and changing the outer loop condition to "while there are frames left to insert" makes it more readable:
```suggestion
  int frames_left = num_frames;
  while (frames_left) {
    if (frame >= sample_nframes) {
      ++sample_id;
      auto sample_shape = input_shape[sample_id];
      std::copy(&sample_shape[first_spatial_dim], &sample_shape[input_shape.sample_dim()],
                frame_shape.begin());
      frame_stride = volume(frame_shape) * type_size;
      if (frame_stride == 0) {
        sample_nframes = 0;
        continue;
      }
      sample_nframes = calc_num_frames(sample_shape, first_spatial_dim);
      data = static_cast<const uint8_t *>(t_list.raw_tensor(sample_id)) + frame_stride * frame_offset;
    }
    tensors.push_back(AsTensor(data, make_span(frame_shape), dtype, nvcv_layout));
    data += frame_stride;
    frame_offset++;
    frames_left--;
 }
```